### PR TITLE
Persona: complete mergeStyles conversion and some small refactoring, add some styling functions.

### LINKED ISSUE #4315: [Persona] getStyles doesn't work
### Bug Report

- __Package version(s)__: Latest

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? No

Requested priority: Normal

Products/sites affected: (if applicable)

#### Describe the issue:

getStyles prop doesn't add styles for Persona component.

### If applicable, please provide a codepen repro:

https://codepen.io/micahgodbolt/pen/LdxVoQ

### PR DESCRIPTION
**Pull request by jordandrako**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #4315
- [x] Include a change request file using `$ npm run change`

### Description of changes

**Convert Persona, PersonaCoin, and PersonaPresence to mergeStyles:**
- Convert sass files to mergeStyles.
- Moved all className and size logic to respective styles files.
- Create `IPersonaSharedProps` that is extended by `IPersonaProps`, `IPersonaCoinProps`, `IPersonaPresenceProps`.
  - Props of type `IRenderFunction<IPersonaProps>` and are used in Persona **AND** PersonaCoin, such as `onRenderCoin`, were changed to `IRenderFunction<IPersonaSharedProps>`.
- Delete sass files
- ~Delete `PersonaConsts.ts`. It was only applying className logic.~
- Refactor `PersonaConsts.ts` to serve size and presence states to styles files.
- Remove unused styles.
  - `rootIsDarkText`, `rootIsSelectable`, `.rootIsReadonly` were in the sass files, but never called from the component, and not accepted in the types.
  - Many classNames replaced with props/conditionals.
- The presence Icon now doesn't render at size 32 or under (either via `PersonaSize` enum or `coinSize <= 32`).

**Added functionality:**
- `noWrap` styles in Styling (Mimics noWrap sass mixin)
- ~`zIndex` namespace in Styling (Mimics `$ms-zIndex` sass variables)~ Should not use zIndex, removed from styles.
- ~Add High Contrast colors to `DefaultPalette` (were missing from sass variable conversion)~

**Other:**
- Update Facepile and ActivityItem to used `IPersonaSharedProps`
- Update Facepile test to `dive()` through decorated Personas.

### Focus areas to test

- [ ] **Test if IPersonaSharedProps breaks contract**
- [x] Test RTL still work correctly.
- [ ] Test if high contrast mode looks accessible.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Persona/Persona.styles.ts [hunk 1]
```diff
@@ -1,16 +1,339 @@
...
+      'ms-Persona-primaryText',
+      noWrap,
+      {
+        color: palette.neutralPrimary,
+        fontWeight: FontWeights.regular,
+        fontSize: FontSizes.large,
+      },
+
+      showSecondaryText && {
+        height: '16px',
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/Persona.styles.ts]: Can we put 16px in a const and reuse elsewhere?

File: packages/office-ui-fabric-react/src/components/Persona/Persona.types.ts [hunk 2]
```diff
@@ -93,64 +84,207 @@ export interface IPersonaProps extends React.HTMLAttributes<PersonaBase> {
...
-  optionalText?: string;
+  coinSize?: number;
+
+  /**
+   * Optional HTML element props for Persona coin.
+   */
+  coinProps?: React.HTMLAttributes<HTMLDivElement>;
+
+  /**
+   * Theme provided by HOC.
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/Persona.types.ts]: Type out "High Order Component" for improved readability

File: packages/office-ui-fabric-react/src/components/Persona/Persona.types.ts [hunk 3]
```diff
@@ -93,64 +84,207 @@ export interface IPersonaProps extends React.HTMLAttributes<PersonaBase> {
...
+   */
+  size?: PersonaSize;
+}
+
+export interface IPersonaCoinStyles {
+  coin: IStyle;
+  imageArea: IStyle;
+  image: IStyle;
+  initials: IStyle;
+  size10NoPresenceIcon: IStyle;
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/Persona.types.ts]: size10WithNoPresenceIcon size10WithoutPresenceIcon is a more natural sounding name

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  // Persona size conditionals
+  const isSize10 = size === PersonaSize.size10 || size === PersonaSize.tiny;
+  const isSize16 = size === PersonaSize.size16;
+  const isSize24 = size === PersonaSize.size24 || size === PersonaSize.extraExtraSmall;
+  const isSize28 = size === PersonaSize.size28 || size === PersonaSize.extraSmall;
+  const isSize32 = size === PersonaSize.size32;
+  const isSize40 = size === PersonaSize.size40 || size === PersonaSize.small;
+  const isSize48 = size === PersonaSize.size48;
+  const isSize72 = size === PersonaSize.size72 || size === PersonaSize.large;
+  const isSize100 = size === PersonaSize.size100 || size === PersonaSize.extraLarge;
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts]: These conditionals are repeated 3 times, we should extract them out to a common place to reduce duplication

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+  // Persona Sizes
+  const personaSize16 = '16px';
+  const personaSize24 = '24px';
+  const personaSize28 = '28px';
+  const personaSize32 = '32px';
+  const personaSize40 = '40px';
+  const personaSize48 = '48px';
+  const personaSize72 = '72px';
+  const personaSize100 = '100px';
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts]: These consts are also repeated several places, so they should probably be extracted to a common place as well

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx [hunk 6]
```diff
@@ -87,37 +81,52 @@ export class PersonaCoin extends React.Component<IPersonaProps, IPersonaState> {
...
     const coinSizeStyle = coinSize ? { width: coinSize, height: coinSize } : undefined;
 
+    const personaPresenceProps: IPersonaPresenceProps = {
+      coinSize,
+      presence,
+      theme,
+      size,
+    };
+
+    const classNames = getClassNames(getStylesProp! || getStyles!, {
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx]: Do we have to use bangs here and elsewhere? I.e. are we absolutely certain getStylesProp/getStyles/theme can never be null? If they can be null, we should avoid using !, to allow typescript to do its nullable type checking

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+  const isSize100 = size === PersonaSize.size100 || size === PersonaSize.extraLarge;
+
+  // Skype presence colors
+  const colorPresenceAvailable = '#7FBA00';
+  const colorPresenceAway = '#FCD116';
+  const colorPresenceBusy = '#E81123';
+  const colorPresenceDndBackground = '#E81123';
+  const colorPresenceDndLine = '#FFFFFF';
+  const colorPresenceOffline = '#93ABBD';
+  // const colorPresenceOutOfOffice = palette.magenta;
```
Comment (mtennoe) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: remove commented out code

File: packages/styling/src/interfaces/IPalette.ts [hunk 8]
```diff
@@ -242,4 +242,21 @@ export interface IPalette {
    * Color code for greenLight.
    */
   greenLight: string;
+
+  /**
+   * Color code for contrast black disabled.
+   */
+  contrastBlackDisabled: string;
```
Comment (mtennoe) [inline packages/styling/src/interfaces/IPalette.ts]: Missing newlines here and below between props

**Comment by mtennoe** [comment]
Can you test that service side rendering of the converted components works as expected as well?

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx (hunk 6 above)
Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx]: I know theme can't be null. It needs to pull the default theme values. I'll test removing the bang from getStyles. I'm sure I put it there in the middle of converting as it was yelling about types, but I probably fixed that by the end.

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts (hunk 5 above)
Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts]: This one is easy enough as a namespace.

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts (hunk 4 above)
Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.styles.ts]: This one, however, is a bit more complicated. I'd have to make it an exported function that takes size as an argument then returns the correct boolean, right?

**Comment by betrue-final-final** [comment]
Hey, the symbols added are great! However they are misaligned. Note the check is low in the circle. 

image

File: packages/styling/src/interfaces/IPalette.ts [hunk 9]
```diff
@@ -242,4 +242,24 @@ export interface IPalette {
    * Color code for greenLight.
    */
   greenLight: string;
+
+  /**
```
Comment (phkuo) [inline packages/styling/src/interfaces/IPalette.ts]: We should ONLY be using system colors for high contrast modes:
https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#System_Colors
@betrue-final-final

File: packages/styling/src/styles/zIndex.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+export namespace zIndex {
```
Comment (phkuo) [inline packages/styling/src/styles/zIndex.ts]: Don't make an API for z-index. We don't want people using z-index.
Additionally this is misleading, it implies that z-index values are global and are all compared relative to each other, where in reality they're only compared against others in their own stacking context:
https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context

We do not want any new instances of z-index. Any old ones should be removed when the opportunity arises. Instead, we should use proper ordering of DOM elements, or use the Callout control for heavier-weight layering needs (like for menus and dialog boxes).

**Comment by jordandrako** [comment]
@betrue-final-final Those icons aren't supposed to be there. Are you reproing that locally or using the screener image?

**Comment by betrue-final-final** [comment]
This is on screener. Why would we not want the icons there? It makes them accessible.

**Comment by jordandrako** [comment]
The icons show up at larger sizes (`size40` and up). If you think I should change that behavior I can, but the images in screener are showing the icon on `size32` where they should have the property `display: none` (and they do locally). So, that seems to be an issue from the way screener renders a story.

**Comment by betrue-final-final** [comment]
I see. Then we should not show them.

**Comment by jordandrako** [comment]
Alright, could you look at why screener is showing that then?

**Comment by jordandrako** [comment]
@betrue-final-final I'm blocked here.
1. Can you take a look at the high contrast color comment above and see if we need to change the previous colors to system colors?
2. How do I debug screener rendering the icon when it doesn't render locally? I could try keeping the icon from rendering at all instead of setting `display: none` at those sizes.

**Comment by betrue-final-final** [comment]
I can look at #1, but I have no idea about #2.

**Comment by lynamemi** [comment]
@jordandrako I'd say trying no. 2 is good call.

File: packages/styling/src/interfaces/IPalette.ts (hunk 8 above)
Comment (jordandrako) [reply packages/styling/src/interfaces/IPalette.ts]: Fixed this, but the comment wasn't marked as outdated.

**Comment by jordandrako** [comment]
@betrue-final-final I fixed the Icon showing when it shouldn't but screener is still showing some weird things. Namely, it's rendering two extra Personas in the PeoplePicker that aren't in the story.

**Comment by betrue-final-final** [comment]
It appears as if you are using system colors, and mapping the "offline", "busy" and "do not disturb" with the disabled color (color_graytext). This seems ok.

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+      {
+        backgroundColor: presenceColorAvailable,
+        position: 'absolute',
+        height: personaPresenceSize.size12,
+        width: personaPresenceSize.size12,
+        borderRadius: '50%',
+        top: 'auto',
+        right: `-${personaPresenceSize.border}`,
+        bottom: `-${personaPresenceSize.border}`,
+        border: `${personaPresenceSize.border} solid ${palette.white}`,
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: use semanticColors.bodyBackground

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            left: 0,
+          },
+
+          [HighContrastSelector]: {
+            color: palette.contrastBlackDisabled,
+            backgroundColor: 'Window',
+
+            selectors: {
+              ':before': {
+                boxShadow: `0 0 0 2px ${palette.contrastBlackDisabled} inset`,
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: @betrue-final-final do we want shadows in high contrast? (do they even render?)
i would prefer just removing (or at least ignoring) shadows, theyre too subtle for high contrast mode

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+              },
+
+              ':after': {
+                backgroundColor: palette.contrastBlackDisabled,
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: use only system colors

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+              }
+            }
+          },
+
+          [HighContrastBWSelector]: {
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: we won't need to deal with every type of HC if we use only system colors

File: packages/styling/src/styles/CommonStyles.ts [hunk 15]
```diff
@@ -1,4 +1,5 @@
 export const HighContrastSelector = '@media screen and (-ms-high-contrast: active)';
+export const HighContrastBWSelector = '@media screen and (-ms-high-contrast: black-on-white)';
```
Comment (phkuo) [inline packages/styling/src/styles/CommonStyles.ts]: we shouldn't need this

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts (hunk 12 above)
Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: This `box-shadow` is needed to make the blocked icon look like this
image

I'll change it to `border` and adjust the height and width to compensate.

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts (hunk 13 above)
Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: @betrue-final-final Can you double check this as it's a visual change? I don't know what system colors I need to be using to make the presence colors differentiate. Maybe on available presence use "Highlight" and "WindowText" for everything else?

Comment (jordandrako) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: **Current High Contrast look:**

image

**Using 'Highlight' on available and 'WindowText' on all else:**

image

**Comment by jordandrako** [comment]
@phkuo I removed the custom contrast colors and went to just system colors. @betrue-final-final and I conversed offline and decided that the presence colors should be `windowtext` unless the presence is available, in which case it should be `highlight`.

I also have some work stashed to help with accessibility and determining presence status (especially in high contrast mode where the colors are now system colors only). Basically, I have the presence component have a `title` (tooltip) with a descriptor/name of the status on hover. I'm thinking that'll have to be saved for another PR though as it defaults to English. I do have a solution for that as well though by making a new `presenceDescriptor` API if the user wants/needs the tooltip to be another language.

File: packages/merge-styles/src/IRawStyleBase.ts [hunk 16]
```diff
@@ -293,6 +293,16 @@ export interface IRawStyleBase extends IRawFontStyle {
    */
   backgroundBlendMode?: ICSSRule | string;
 
+  /**
+   * The background-clip CSS property specifies if an element's
+   * background, whether a <color> or an <image>, extends
+   * underneath its border.
+   *
+   * The `text` value is experimental and should not be used
```
Comment (phkuo) [inline packages/merge-styles/src/IRawStyleBase.ts]: should also note it doesn't work in IE

File: packages/styling/src/styles/GeneralStyles.ts [hunk 17]
```diff
@@ -7,4 +7,11 @@ export const normalize: IRawStyle = {
   margin: 0,
   padding: 0,
   boxSizing: 'border-box'
-};
\ No newline at end of file
+};
+
+export const noWrap: IRawStyle = {
+  display: 'block',
```
Comment (phkuo) [inline packages/styling/src/styles/GeneralStyles.ts]: probably would prefer inline-block here, so things don't take up all available width
although i'd prefer just leaving off display altogether and letting the consumer decide between block/inline-block

Comment (jordandrako) [reply packages/styling/src/styles/GeneralStyles.ts]: Probably a good idea. Thanks for all your thorough reviewing!

File: packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx [hunk 18]
```diff
@@ -85,80 +77,106 @@ export class PersonaCoin extends React.Component<IPersonaProps, IPersonaState> {
...
 
+    const personaPresenceProps: IPersonaPresenceProps = {
+      coinSize,
+      presence,
+      size,
+      theme,
+    };
+
+    // Use getStyles from props, or fall back to getStyles from styles file.
+    const classNames = getClassNames(getStylesProp || getStyles, {
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Persona/PersonaCoin.tsx]: The `styled` function already combines getStyles from props! :)

File: packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+              top: '50%',
+              left: 0,
+            } : undefined,
+
+            [HighContrastSelector]: {
+              backgroundColor: 'WindowText',
+
+              selectors: {
+                ':before': {
+                  width: `calc(100% - ${personaPresenceSize.border})`,
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Persona/PersonaPresence.styles.ts]: really? Can we avoid calc? (for reflow reasons?)

**Pull request merged**